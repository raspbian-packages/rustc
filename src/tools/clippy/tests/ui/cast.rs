#![feature(plugin)]
#![plugin(clippy)]

#[warn(cast_precision_loss, cast_possible_truncation, cast_sign_loss, cast_possible_wrap, cast_lossless)]
#[allow(no_effect, unnecessary_operation)]
fn main() {
    // Test cast_precision_loss
    1i32 as f32;
    1i64 as f32;
    1i64 as f64;
    1u32 as f32;
    1u64 as f32;
    1u64 as f64;
    // Test cast_possible_truncation
    1f32 as i32;
    1f32 as u32;
    1f64 as f32;
    1i32 as i8;
    1i32 as u8;
    1f64 as isize;
    1f64 as usize;
    // Test cast_possible_wrap
    1u8 as i8;
    1u16 as i16;
    1u32 as i32;
    1u64 as i64;
    1usize as isize;
    // Test cast_lossless with casts to integer types
    1i8 as i16;
    1i8 as i32;
    1i8 as i64;
    1u8 as i16;
    1u8 as i32;
    1u8 as i64;
    1u8 as u16;
    1u8 as u32;
    1u8 as u64;
    1i16 as i32;
    1i16 as i64;
    1u16 as i32;
    1u16 as i64;
    1u16 as u32;
    1u16 as u64;
    1i32 as i64;
    1u32 as i64;
    1u32 as u64;
    // Test cast_lossless with casts to floating-point types
    1i8 as f32;
    1i8 as f64;
    1u8 as f32;
    1u8 as f64;
    1i16 as f32;
    1i16 as f64;
    1u16 as f32;
    1u16 as f64;
    1i32 as f64;
    1u32 as f64;
    // Test cast_lossless with casts from floating-point types
    1.0f32 as f64;
    // Test cast_sign_loss
    1i32 as u32;
    1isize as usize;
    // Extra checks for *size
    // Casting from *size
    1isize as i8;
    1isize as f64;
    1usize as f64;
    1isize as f32;
    1usize as f32;
    1isize as i32;
    1isize as u32;
    1usize as u32;
    1usize as i32;
    // Casting to *size
    1i64 as isize;
    1i64 as usize;
    1u64 as isize;
    1u64 as usize;
    1u32 as isize;
    1u32 as usize; // Should not trigger any lint
    1i32 as isize; // Neither should this
    1i32 as usize;
    // Test cast_unnecessary
    1i32 as i32;
    1f32 as f32;
    false as bool;
    &1i32 as &i32;
    // Should not trigger
    let v = vec!(1);
    &v as &[i32];
    1.0 as f64;
    1 as u64;
}
