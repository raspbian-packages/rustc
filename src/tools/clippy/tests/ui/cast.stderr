error: casting i32 to f32 causes a loss of precision (i32 is 32 bits wide, but f32's mantissa is only 23 bits wide)
 --> $DIR/cast.rs:8:5
  |
8 |     1i32 as f32;
  |     ^^^^^^^^^^^
  |
  = note: `-D cast-precision-loss` implied by `-D warnings`

error: casting i64 to f32 causes a loss of precision (i64 is 64 bits wide, but f32's mantissa is only 23 bits wide)
 --> $DIR/cast.rs:9:5
  |
9 |     1i64 as f32;
  |     ^^^^^^^^^^^

error: casting i64 to f64 causes a loss of precision (i64 is 64 bits wide, but f64's mantissa is only 52 bits wide)
  --> $DIR/cast.rs:10:5
   |
10 |     1i64 as f64;
   |     ^^^^^^^^^^^

error: casting u32 to f32 causes a loss of precision (u32 is 32 bits wide, but f32's mantissa is only 23 bits wide)
  --> $DIR/cast.rs:11:5
   |
11 |     1u32 as f32;
   |     ^^^^^^^^^^^

error: casting u64 to f32 causes a loss of precision (u64 is 64 bits wide, but f32's mantissa is only 23 bits wide)
  --> $DIR/cast.rs:12:5
   |
12 |     1u64 as f32;
   |     ^^^^^^^^^^^

error: casting u64 to f64 causes a loss of precision (u64 is 64 bits wide, but f64's mantissa is only 52 bits wide)
  --> $DIR/cast.rs:13:5
   |
13 |     1u64 as f64;
   |     ^^^^^^^^^^^

error: casting f32 to i32 may truncate the value
  --> $DIR/cast.rs:15:5
   |
15 |     1f32 as i32;
   |     ^^^^^^^^^^^
   |
   = note: `-D cast-possible-truncation` implied by `-D warnings`

error: casting f32 to u32 may truncate the value
  --> $DIR/cast.rs:16:5
   |
16 |     1f32 as u32;
   |     ^^^^^^^^^^^

error: casting f32 to u32 may lose the sign of the value
  --> $DIR/cast.rs:16:5
   |
16 |     1f32 as u32;
   |     ^^^^^^^^^^^
   |
   = note: `-D cast-sign-loss` implied by `-D warnings`

error: casting f64 to f32 may truncate the value
  --> $DIR/cast.rs:17:5
   |
17 |     1f64 as f32;
   |     ^^^^^^^^^^^

error: casting i32 to i8 may truncate the value
  --> $DIR/cast.rs:18:5
   |
18 |     1i32 as i8;
   |     ^^^^^^^^^^

error: casting i32 to u8 may lose the sign of the value
  --> $DIR/cast.rs:19:5
   |
19 |     1i32 as u8;
   |     ^^^^^^^^^^

error: casting i32 to u8 may truncate the value
  --> $DIR/cast.rs:19:5
   |
19 |     1i32 as u8;
   |     ^^^^^^^^^^

error: casting f64 to isize may truncate the value
  --> $DIR/cast.rs:20:5
   |
20 |     1f64 as isize;
   |     ^^^^^^^^^^^^^

error: casting f64 to usize may truncate the value
  --> $DIR/cast.rs:21:5
   |
21 |     1f64 as usize;
   |     ^^^^^^^^^^^^^

error: casting f64 to usize may lose the sign of the value
  --> $DIR/cast.rs:21:5
   |
21 |     1f64 as usize;
   |     ^^^^^^^^^^^^^

error: casting u8 to i8 may wrap around the value
  --> $DIR/cast.rs:23:5
   |
23 |     1u8 as i8;
   |     ^^^^^^^^^
   |
   = note: `-D cast-possible-wrap` implied by `-D warnings`

error: casting u16 to i16 may wrap around the value
  --> $DIR/cast.rs:24:5
   |
24 |     1u16 as i16;
   |     ^^^^^^^^^^^

error: casting u32 to i32 may wrap around the value
  --> $DIR/cast.rs:25:5
   |
25 |     1u32 as i32;
   |     ^^^^^^^^^^^

error: casting u64 to i64 may wrap around the value
  --> $DIR/cast.rs:26:5
   |
26 |     1u64 as i64;
   |     ^^^^^^^^^^^

error: casting usize to isize may wrap around the value
  --> $DIR/cast.rs:27:5
   |
27 |     1usize as isize;
   |     ^^^^^^^^^^^^^^^

error: casting f32 to f64 may become silently lossy if types change
  --> $DIR/cast.rs:29:5
   |
29 |     1.0f32 as f64;
   |     ^^^^^^^^^^^^^ help: try: `f64::from(1.0f32)`
   |
   = note: `-D cast-lossless` implied by `-D warnings`

error: casting u8 to u16 may become silently lossy if types change
  --> $DIR/cast.rs:31:5
   |
31 |     (1u8 + 1u8) as u16;
   |     ^^^^^^^^^^^^^^^^^^ help: try: `u16::from(1u8 + 1u8)`

error: casting i32 to u32 may lose the sign of the value
  --> $DIR/cast.rs:33:5
   |
33 |     1i32 as u32;
   |     ^^^^^^^^^^^

error: casting isize to usize may lose the sign of the value
  --> $DIR/cast.rs:34:5
   |
34 |     1isize as usize;
   |     ^^^^^^^^^^^^^^^

error: casting to the same type is unnecessary (`i32` -> `i32`)
  --> $DIR/cast.rs:37:5
   |
37 |     1i32 as i32;
   |     ^^^^^^^^^^^
   |
   = note: `-D unnecessary-cast` implied by `-D warnings`

error: casting to the same type is unnecessary (`f32` -> `f32`)
  --> $DIR/cast.rs:38:5
   |
38 |     1f32 as f32;
   |     ^^^^^^^^^^^

error: casting to the same type is unnecessary (`bool` -> `bool`)
  --> $DIR/cast.rs:39:5
   |
39 |     false as bool;
   |     ^^^^^^^^^^^^^

error: aborting due to 28 previous errors

